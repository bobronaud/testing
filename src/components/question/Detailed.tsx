const Detailed = () => {
  return <></>;
};

export default Detailed;
