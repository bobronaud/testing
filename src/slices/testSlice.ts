import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from '../app/store';

export type Question = {
  type: string;
  question: string;
  answers: string[] | null;
  rightAnswer: string | string[] | null;
};

type Answer = Question & {
  userAnswer: string;
};

type Phase = 'starting' | 'inProcess' | 'finished';

type Completed = {
  [key: number]: boolean;
};

interface state {
  phase: Phase;
  questions: Question[];
  currentQuestion: Question | null;
  currentStep: number;
  completedSteps: Completed;
  result: Answer[];
}
const initialState: state = {
  phase: 'starting',
  questions: [],
  currentQuestion: null,
  currentStep: 0,
  completedSteps: {},
  result: [],
};

const testSlice = createSlice({
  name: 'test',
  initialState,
  reducers: {
    writeAnswer: (state, action: PayloadAction<string | string[]>) => {
      const userAnswer = action.payload;
      const answer = Object.assign(state.currentQuestion as Question, {
        userAnswer,
      }) as Answer;
      state.result = [...state.result, answer];
    },
    finishStep: (state) => {
      const isLastStep = state.currentStep + 1 === state.questions.length;
      if (isLastStep) {
        state.phase = 'finished';
      } else {
        state.completedSteps[state.currentStep] = true;
        state.currentStep += 1;
        state.currentQuestion = state.questions[state.currentStep];
      }
    },
    setPhase: (state, action: PayloadAction<Phase>) => {
      const newPhase = action.payload;
      state.phase = newPhase;
    },
    uploadQuestions: (state, action: PayloadAction<Question[]>) => {
      const questions = action.payload;
      state.questions = questions;
      state.currentQuestion = questions[0];
    },
  },
});

export default testSlice.reducer;
export const selectCards = (state: RootState) => state.test;
export const { writeAnswer, finishStep, setPhase, uploadQuestions } = testSlice.actions;
