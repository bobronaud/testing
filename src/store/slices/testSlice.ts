import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from '../store';

export type Question = {
  type: string;
  question: string;
  answers: string[] | null;
  rightAnswer: string | string[] | null;
};

type Answer = Question & {
  userAnswer: string;
};

export type Phase = 'starting' | 'inProcess' | 'finished';

type Completed = {
  [key: number]: boolean;
};

export type State = {
  timer: boolean;
  time: [number, number];
  phase: Phase | null;
  questions: Question[];
  currentQuestion: Question | null;
  currentStep: number;
  completedSteps: Completed;
  result: Answer[];
};
const initialState: State = {
  timer: true,
  time: [5, 0],
  phase: null,
  questions: [],
  currentQuestion: null,
  currentStep: 0,
  completedSteps: {},
  result: [],
};

const testSlice = createSlice({
  name: 'test',
  initialState,
  reducers: {
    writeAnswer: (state, action: PayloadAction<string | string[]>) => {
      const userAnswer = action.payload;
      const answer = Object.assign(state.currentQuestion as Question, {
        userAnswer,
      }) as Answer;
      state.result = [...state.result, answer];
    },
    finishStep: (state) => {
      const isLastStep = state.currentStep + 1 === state.questions.length;
      if (isLastStep) {
        state.phase = 'finished';
      } else {
        state.completedSteps[state.currentStep] = true;
        state.currentStep += 1;
        state.currentQuestion = state.questions[state.currentStep];
      }
    },
    setPhase: (state, action: PayloadAction<Phase>) => {
      const newPhase = action.payload;
      state.phase = newPhase;
    },
    setTimer: (state, action: PayloadAction<boolean>) => {
      const value = action.payload;
      state.timer = value;
    },
    setTime: (state, action: PayloadAction<[number, number]>) => {
      const time = action.payload;
      state.time = time;
    },
    uploadQuestions: (state, action: PayloadAction<Question[]>) => {
      const questions = action.payload;
      state.questions = questions;
      state.currentQuestion = questions[0];
    },
    setDataApp: (state, action: PayloadAction<State>) => {
      const data = action.payload;
      const newState = Object.assign(state, data);
      state = newState;
    },
    resetProgress: (state) => {
      const newState = Object.assign(state, initialState);
      state = newState;
    },
  },
});

export default testSlice.reducer;
export const selectCards = (state: RootState) => state.test;
export const {
  writeAnswer,
  finishStep,
  setPhase,
  setTimer,
  setTime,
  setDataApp,
  uploadQuestions,
  resetProgress,
} = testSlice.actions;
