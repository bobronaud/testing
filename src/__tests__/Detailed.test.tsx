import userEvent from '@testing-library/user-event';
import Detailed from '../components/question/Detailed';
import { State } from '../store/slices/testSlice';
import { chooseOne, detailed } from './__fixtures__/questionsData';
import renderWithRedux from './helpers/renderWithRedux';
import { act } from 'react-dom/test-utils';

describe('тест компонента Detailed', () => {
  const initialState = {
    test: {
      timer: false,
      time: [5, 0],
      phase: 'inProcess',
      questions: [detailed, chooseOne],
      currentQuestion: detailed,
      currentStep: 0,
      completedSteps: {},
      result: [],
    } as State,
  };
  test('отрисовка', () => {
    const container = renderWithRedux(<Detailed />, initialState);
    expect(container).toMatchSnapshot();
  });

  test('работа инпута', async () => {
    const { queryByText, getByText, getByRole } = renderWithRedux(
      <Detailed />,
      initialState,
    );
    const input = getByRole('textbox');
    expect(input).toHaveFocus();

    const btn = getByRole('button');
    await act(async () => {
      await userEvent.click(btn);
    });
    const errorText = getByText(/Необходимо ввести ответ/i);
    expect(errorText).toBeInTheDocument();

    await act(async () => {
      await userEvent.type(input, '\n      ');
    });
    await act(async () => {
      await userEvent.click(btn);
    });
    expect(errorText).toBeInTheDocument();

    await act(async () => {
      await userEvent.clear(input);
    });
    await act(async () => {
      await userEvent.type(input, 'не знаю');
    });
    const error = queryByText(/Необходимо ввести ответ/i);
    expect(error).not.toBeInTheDocument();
  });

  test('работоспособность сабмита', async () => {
    const { queryByText, getByText, getByRole } = renderWithRedux(
      <Detailed />,
      initialState,
    );
    const input = getByRole('textbox');
    const btn = getByRole('button');
    await act(async () => {
      await userEvent.type(input, 'правильный ответ');
    });
    await act(async () => {
      await userEvent.click(btn);
    });
    const prevQuestion = queryByText(detailed.question);
    const nextQuestion = getByText(chooseOne.question);
    expect(prevQuestion).not.toBeInTheDocument();
    expect(nextQuestion).toBeInTheDocument();
  });
});
