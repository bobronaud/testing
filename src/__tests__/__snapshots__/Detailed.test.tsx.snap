// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`тест компонента Detailed отрисовка 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root css-0"
      >
        <h3
          class="MuiTypography-root MuiTypography-h6 css-1iz74ne-MuiTypography-root"
        >
          Опишите кратко процесс фотосинтеза.
        </h3>
        <div
          class="MuiFormControl-root css-q8hpuo-MuiFormControl-root"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl MuiInputBase-multiline css-8ewcdo-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <textarea
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r0:"
                rows="5"
                style="height: 0px; overflow: hidden;"
              />
              <textarea
                aria-hidden="true"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                readonly=""
                style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"
                tabindex="-1"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-ihdtdm"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <p
            class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained css-1wc848c-MuiFormHelperText-root"
          />
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1eafrsi-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Ответить
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-0"
    >
      <h3
        class="MuiTypography-root MuiTypography-h6 css-1iz74ne-MuiTypography-root"
      >
        Опишите кратко процесс фотосинтеза.
      </h3>
      <div
        class="MuiFormControl-root css-q8hpuo-MuiFormControl-root"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl MuiInputBase-multiline css-8ewcdo-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <textarea
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r0:"
              rows="5"
              style="height: 0px; overflow: hidden;"
            />
            <textarea
              aria-hidden="true"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
              readonly=""
              style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"
              tabindex="-1"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-ihdtdm"
              >
                <span
                  class="notranslate"
                >
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <p
          class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained css-1wc848c-MuiFormHelperText-root"
        />
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1eafrsi-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Ответить
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
